rococo-runtime: `RococoGenesisExt` removed

michalkucharczyk opened the pull request:
[`RococoGenesisExt`](https://github.com/paritytech/polkadot-sdk/blob/a414ea7515c9cdc81f1d12410e646afc148250e8/polkadot/node/service/src/chain_spec.rs#L152-L171) is removed. It was the hack to allow overwriting `EpochDurationInBlocks`. Removal of `RococGenesisExt` prevents from manipulating the state to change the runtime constants.

Changes:
- Environment variable which controls the `time::EpochDurationInBlocks` value was added: `ROCOCO_EPOCH_DURATION` (epoch duration will be set to the value of env),
- `10,100,600` versions of rococo-runtime are built in CI and put into `polkadot-debug` docker image.

`rococo-runtime` building examples:
- to build runtime for `versi_staging_testnet` which had  EpochDurationInBlocks set to 100:
  ```
  ROCOCO_EPOCH_DURATION=100 cargo build --features=fast-runtime -p rococo-runtime
  ```
- to build runtime for `wococo_development`
  ```
  ROCOCO_EPOCH_DURATION=10 cargo build --features=fast-runtime -p rococo-runtime
  ```
- to build `versi-staging` chain spec:
  ```
  ROCOCO_EPOCH_DURATION=100 cargo run -p polkadot --features=fast-runtime -- build-spec --chain versi-staging --raw 
  ```
- to build `wococo-dev` chain spec:
  ```
  ROCOCO_EPOCH_DURATION=10 cargo run -p polkadot  --features=fast-runtime  -- build-spec --chain wococo-dev --raw 
  ```

Chain-specs can also be built using [provided sample script](https://github.com/paritytech/polkadot-sdk/blob/d55ae688a8d0769966a125fde2f0dd6cc7f8e762/polkadot/scripts/update-chain-spec-with-fast-rococo.sh). As the epoch duration is hard-coded into wasm blob, it is enough to replace the `code` field.

Step towards: paritytech/polkadot-sdk#25

michalkucharczyk commented:
It is part extracted of https://github.com/paritytech/polkadot-sdk/pull/1256

michalkucharczyk commented:
@paritytech/devrel  , @the-right-joyce

I wanted to give you a heads-up about some changes that have just been made in this PR. I've modified the default `EpochDurationInBlocks` for several `ChainSpecs`, namely: `versi-local`, `wococo-local`, `rococo-local`, wococo-dev`, `versi-dev`, `rococo-dev` (where duration was previously set to 10), and `versi-staging` (where duration was previously set to 100).

You can find additional details in the PR description.

This change could affect tutorials (and docs maybe?) that rely on timing within these runtimes. Therefore, I request you to please review and update them accordingly, and include information on how to adjust timings based on these changes.

Thank you for your attention.

_note:_ this is [copy of comment](https://github.com/paritytech/polkadot/pull/7508#issuecomment-1665128112) previously made in `polkadot` repo.

bkchr commented on polkadot/runtime/rococo/build.rs:
```diff
@@ -14,12 +14,22 @@
+
+	if env::var(ROCOCO_EPOCH_DURATION_ENV).is_ok() && env::var("CARGO_FEATURE_FAST_RUNTIME").is_ok()
+	{
+		builder = builder.enable_feature("fast-runtime")
```
The wasm builder automatically detects enabled features and passes them to the wasm build.

ordian commented:
cc @Sophia-Gold since you're using `session_length_in_blocks` in your test

PierreBesson commented:
From the point of view of the Devops team that is generating those chainspec to launch testnets, this doesn't look  very 
practical to use as it will force us to work from source instead from a binary for just 1 config value tweak.

If we could simply have a `--chain rococo-fast-runtime` already available in the built node it would be most convenient to do our job and automate our processes for launching custom Rococo-based testnet such as Versi.

bkchr commented:
> From the point of view of the Devops team that is generating those chainspec to launch testnets, this doesn't look very
> practical to use as it will force us to work from source instead from a binary for just 1 config value tweak.

Our test nets are running with at least 1 hour session length any way or not? And 1 hour is still the default.

eskimor commented:
> > From the point of view of the Devops team that is generating those chainspec to launch testnets, this doesn't look very
> > practical to use as it will force us to work from source instead from a binary for just 1 config value tweak.
> 
> Our test nets are running with at least 1 hour session length any way or not? And 1 hour is still the default.

No, Versi has shorter sessions and we would like to keep it at that as session changes are something very crucial to test.

bkchr commented:
> No, Versi has shorter sessions and we would like to keep it at that as session changes are something very crucial to test.

Yeah and it will still be possible. It will just work differently. This entire changing of the session length was a hack from the beginning, that I should never have used :P

PierreBesson commented:
I appreciate the change to make it fully configurable. However, being able to change runtime constants without building from source could be very useful. But I guess, we will have to accept that we won't have a way to build fast-runtime chainspecs from the released binary after this change.

@michalkucharczyk
- Would it be possible to keep the short default epoch duration for the Versi runtime ?
- Would it also be possible to offer a configurable epoch duration for the westend runtime as well ?

michalkucharczyk commented:
>  However, being able to change runtime constants without building from source could be very useful.

We want to go away from configuring runtime parameters using _state trie_. 

The only other solution I see, is to - by default - build the rococo runtime binary with smaller `epochDuration` value and name it name it `versi-runtime`. And hide it behind the feature as other runtimes. This will lengthen the build process.
**If that is acceptable, I can do it.** However I believe we should not aim into adding more pre-configured runtimes into default build. Adding versi seems fair to me.

The following values were used in runtimes, I assume that we are talking only about keeping `staging` versions?:
|  | dev | local | staging
--- | --- | --- | ---
rococo  | 10 | 10 | 600
versi | 10 | 10 | 100
wococo | 10 | 10 | 


> Would it also be possible to offer a configurable epoch duration for the westend runtime as well ?
Technically should be doable. The question is if we should build another runtime version by default or is it OK for you to build it on your side? Maybe we should move this request to new ticket?

bkchr commented:
> I appreciate the change to make it fully configurable. However, being able to change runtime constants without building from source could be very useful.

This is one of the only ones we support this. I have done this as a hack in 2020 or somewhere then to make it easy to test stuff locally. Versi for sure doesn't fall into this category and I see no problem for you to just rebuild the runtime. In the future we will also be able to support to only build the runtime to build a chain spec, so you will not need to build the entire node. This will then result in a quite fast turn around (assuming you only change something in the runtime `lib.rs`).

This said, we probably should take the performance hit and also build a fast runtime for `rococo-local` for now. Sadly parachains are still not able to onboard at genesis and thus, people would be confused why stuff isn't working as they expect it to work.

bkchr commented:
The native runtime isn't used anymore. So, this is fine.

We can do something like [here](https://github.com/paritytech/substrate/blob/e526f80a91f75a3811059d4ee19b0f190005c5a2/test-utils/runtime/build.rs#L21-L56).

michalkucharczyk commented:
> The native runtime isn't used anymore. So, this is fine.

It is, for chain spec (genesis config).

We need to merge it as part of: https://github.com/paritytech/polkadot-sdk/pull/1256

bkchr commented:
Yeah no, we will not support three different lengths for the epoch duration.

@michalkucharczyk can provide you a script for building the stuff in the correct way.

@BulatSaif given that you already patch the chainspec manually, you could also overwrite the `system.code` field. Then you would only need to recompile the `runtime` which is way faster.

BulatSaif commented:
> @BulatSaif given that you already patch the chainspec manually, you could also overwrite the system.code field. Then you would only need to recompile the runtime which is way faster.

We did it in the past as a result we had a chain that is impossible to update (current wococo). After we apply the runtime update chain will stop producing the blocks `Error Corrupted state at`.  
I assume some migration depends on session length, or runtime which I built was incorrect: https://github.com/paritytech/polkadot/pull/7416 

I would like to avoid modifying the substrate code by the DevOps team as we are not familiar with the codebase. 

But we need a way to deploy and update chains with small session time. It will speed up the development of the parachain team since they can onboard or off board parachains faster.

michalkucharczyk commented:
> @michalkucharczyk can provide you a script for building the stuff in the correct way.

https://github.com/paritytech/polkadot-sdk/pull/1490/commits/f84cd5d1e25c7c9ccd420b7d8641ee2d7249df9e

BulatSaif commented:
>  You will not need to modify any code, you will just need to set an env variable at build time to the appropriate number.

This is the good news, can we have similar option `ROCOCO_EPOCH_DURATION` for westend?

michalkucharczyk commented:
I did some experiment, and added pre-built runtimes into [`polkadot-debug`](https://hub.docker.com/r/paritypr/polkadot-debug/tags) docker image under `/polkadot/runtimes` directory.

Rutimes are not explicitly built by `cargo build ...` command. I added some extra commands into CI in order to build 10 and 100 versions of rococo runtime. 

Runtimes can be extracted with somewhat ugly command:
`docker export $(docker create docker.io/paritypr/polkadot-debug:1490-657ae27f) | tar --transform='flags=r;s|polkadot/runtimes/|runtimes/|' -xf - polkadot/runtimes/rococo-runtime-10/rococo_runtime.wasm`
and then used with the provided [script](https://github.com/paritytech/polkadot-sdk/blob/657ae27ffd5cd853a7c83841d65ea10fd69af25c/polkadot/scripts/build-fast-rococo-and-update-chain-spec.sh) to generate chain-specs.

If that is acceptable, we could keep this as final solution. 
If not, we could merge it, and keep it until devops team finds out how to deal with change introduced by this PR.

BulatSaif commented:
> If that is acceptable, we could keep this as final solution.
> If not, we could merge it, and keep it until devops team finds out how to deal with change introduced by this PR.

The proposed alternative solution to build chainspec with shorter epoch duration looks good. I did not run the scripts but it looks clear and we can use it.

bkchr commented on polkadot/runtime/rococo/build.rs:
```diff
@@ -14,12 +14,21 @@
+// prod_or_fast macro.
+const ROCOCO_EPOCH_DURATION_ENV: &str = "ROCOCO_EPOCH_DURATION";
+
 fn main() {
```
```suggestion
[cfg(feature = "std")]
fn main() {
```

bkchr commented on polkadot/runtime/rococo/build.rs:
```diff
@@ -14,12 +14,21 @@
+	builder.build();
+
+	println!("cargo:rerun-if-env-changed={}", ROCOCO_EPOCH_DURATION_ENV);
 }
```
```suggestion
}

#[cfg(not(feature = "std"))]
fn main() {}
```

bkchr replied:
Please also make `substrate-wasm-builder` an optional dep. (This is done to speedup the build process)

michalkucharczyk replied:
done in d55ae688a8

bkchr commented on polkadot/scripts/build-fast-rococo-and-update-chain-spec.sh:
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash -eux
```
```suggestion
#!/usr/bin/env bash -eux
```

bkchr replied:
Some docs on what these files are doing would be nice.

bkchr commented on polkadot/scripts/update-chain-spec-with-fast-rococo.sh:
```diff
@@ -0,0 +1,63 @@
+#!/bin/bash -eu
```
```suggestion
#!/usr/bin/env bash -eu
```

bkchr commented on polkadot/scripts/build-fast-rococo-and-update-chain-spec.sh:
```diff
@@ -0,0 +1,37 @@
+EPOCH_DURATION_IN_BLOCKS=10
+
+# polkadot command:
+DOCKER_IMAGE=paritypr/polkadot-debug:1539-2085d3f0
```
You are hard coding an image? Sounds like a bad idea?

bkchr commented on polkadot/runtime/rococo/constants/src/lib.rs:
```diff
@@ -41,9 +41,9 @@ pub mod time {
+
 	frame_support::parameter_types! {
-		pub storage EpochDurationInBlocks: BlockNumber = DEFAULT_EPOCH_DURATION;
+		pub storage EpochDurationInBlocks: BlockNumber = prod_or_fast!(1 * HOURS, 1 * MINUTES, "ROCOCO_EPOCH_DURATION");
```
```suggestion
		pub storage EpochDurationInBlocks: BlockNumber = option_env!("ROCOCO_EPOCH_DURATION").map(|s| s.parse().expect("`ROCOCO_EPOCH_DURATION` is not a valid `BlockNumber`")).unwrap_or(1 * MINUTES);
```

Let's take `1 Minute` as default.

michalkucharczyk replied:
@bkchr one question: `prod_or_fast` is also used in other rococo parameters, e.g:
https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/runtime/rococo/src/lib.rs#L382-L387

If we read env directly, then we don't require to have `fast-runtime` feature enabled.

Question is: when changing epoch duration shall we also pass `fast-runtime` feature? Not sure if values of all those parameters shall be kept in sync somehow.

bkchr replied:
I mean we may can change this hear and also make it dependent on the `fast-runtime` feature. My "reasoning" here more was that I would take 1 minute as the default value and there would exist no "faster default value". However, we could still make the env only work with the feature being enabled.

michalkucharczyk replied:
Seems all `prod_or_fast` params in rococo runtime are now gone. So this is not a problem anymore.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3799501

michalkucharczyk commented on polkadot/scripts/update-chain-spec-with-fast-rococo.sh:
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env bash -eu
```
@PierreBesson @BulatSaif 
Will this script be useful for you?
Are you going to directly use it, or could it be removed?

Also question about pre-built rococo runtime delivered in docker image. Would that be useful for you?

PierreBesson replied:
Yes this script is indeed very useful. Thank you. However it doesn't have to be in this repo as it is quite a special use case. I think it would be enough to document the `ROCOCO_EPOCH_DURATION` variable in a readme file.

michalkucharczyk replied:
removed in: [4325737](https://github.com/paritytech/polkadot-sdk/pull/1490/commits/43257373908a53c1755a69f82410830235c36807)
